DirectProgramming/DPC++FPGA/Tutorials/Tools/experimental/platform-designer/add-oneapi/src/add.cpp

```diff
@@ -0,0 +1,133 @@
+#include <iostream>
+
+// oneAPI headers
+#include <sycl.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/ext/intel/prototype/host_pipes.hpp>
+
+// use host pipes to write into addresses in the CSR
+using OutputPipe = sycl::ext::intel::prototype::pipe<
+    class ID_PipeOut, int, 1,
+    // choose defaults for these 4:
+    0, 1, true, false,
+    // store the most recently processed index
+    sycl::ext::intel::prototype::internal::protocol_name::AVALON_MM>;
+
+class Add_Kernel {
+ public:
+  int a;
+  int b;
+
+  void operator()() const {
+    int sum = a + b;
+
+    OutputPipe::write(sum);
+  }
+};
+
+// use host pipes to read from addresses in the CSR
+using InputPipeA = sycl::ext::intel::prototype::pipe<
+    class ID_A, int, 1,
+    // choose defaults for these 4:
+    0, 1, true, false,
+    // store the most recently processed index
+    sycl::ext::intel::prototype::internal::protocol_name::AVALON_MM>;
+using InputPipeB = sycl::ext::intel::prototype::pipe<
+    class ID_B, int, 1,
+    // choose defaults for these 4:
+    0, 1, true, false,
+    // store the most recently processed index
+    sycl::ext::intel::prototype::internal::protocol_name::AVALON_MM>;
+
+// use host pipes to write into addresses in the CSR
+using OutputPipeC = sycl::ext::intel::prototype::pipe<
+    class ID_C, int, 1,
+    // choose defaults for these 4:
+    0, 1, true, false,
+    // store the most recently processed index
+    sycl::ext::intel::prototype::internal::protocol_name::AVALON_MM>;
+
+class AddCSRPipes_Kernel {
+ public:
+  void operator()() const {
+    int a = InputPipeA::read();
+    int b = InputPipeB::read();
+
+    int sum = a + b;
+
+    OutputPipeC::write(sum);
+  }
+};
+
+int main() {
+  bool passed = false;
+
+  try {
+// choose a selector that was selected by the default FPGA build system.
+#if FPGA_SIMULATOR
+    std::cout << "using FPGA Simulator." << std::endl;
+    sycl::queue q(sycl::ext::intel::fpga_simulator_selector{});
+#elif FPGA_HARDWARE
+    std::cout << "using FPGA Hardware." << std::endl;
+    sycl::queue q(sycl::ext::intel::fpga_selector{});
+#else  // #if FPGA_EMULATOR
+    std::cout << "using FPGA Emulator." << std::endl;
+    sycl::queue q(sycl::ext::intel::fpga_emulator_selector{});
+#endif
+
+    int a = 3;
+    int b = 76;
+
+    int expectedSum = a + b;
+
+    std::cout << "add two integers using CSR for input." << std::endl;
+
+    q.single_task<class Add>(Add_Kernel{a, b}).wait();
```

Comment:
Is this wait required? The pipe is going to be blocking at line 91

Reply:
good catch. in this case it's ok because the pipe is only written to once and it has a capacity of 1, but in general that can cause hangs.